Release script should stop on error

<!-- First, thank you for contributing.
Please provide us with at least the following data.

# Prerequisites

Please answer the following questions for yourself before submitting an issue. **YOU MAY DELETE THE PREREQUISITES SECTION.**

- [ ] I am running the latest version
- [ ] I checked the documentation and found no answer
- [ ] I checked to make sure that this issue has not already been filed
-->
## Are you reporting a bug or a feature?

 - [x] Bug
 - [ ] Feature

## Expected Behavior

<!-- Please describe the behavior you are expecting. -->

Should stop on error.

## Actual Behavior

<!-- What is the current behavior? -->

Continues on error.

## Steps to Reproduce the Problem

<!-- What do we need to do, to reproduce your issue?
Please provide screencasts and/or screenshots - awesome tools are:
- MacOS: [GIPHY Capture](https://giphy.com/apps/giphycapture)
- Windows: [ScreenToGif](http://www.screentogif.com/) -->

1. `npm run release`
 
## Failure Logs

<!-- Please include any relevant log snippets or files here. -->

```sh
Step 2 complete...
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @axa-ch/patterns-library@2.0.1-beta.144 test-eslint: `eslint src/`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the @axa-ch/patterns-library@2.0.1-beta.144 test-eslint script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/axawinterthur/.npm/_logs/2018-06-14T12_32_10_561Z-debug.log
Step 3 complete...
Switched to branch 'develop'
To github.com:axa-ch/patterns-library.git
   505240e..dc5a247  develop -> develop
To github.com:axa-ch/patterns-library.git
 * [new tag]         v2.0.1-beta.144 -> v2.0.1-beta.144
 * [new tag]         v2.0.1-beta.145 -> v2.0.1-beta.145
Step 4 complete...
Switched to branch 'master'
To github.com:axa-ch/patterns-library.git
   9f9ab4b..fbc6f01  master -> master
Everything up-to-date
Step 5 complete...
Switched to branch 'develop'
To github.com:axa-ch/patterns-library.git
   dc5a247..fbc6f01  develop -> develop
Everything up-to-date
Step 6 complete! Publishing done successfully. Have fun!

Already on 'develop'
```

--- a/stack/tasks/release.js
+++ b/stack/tasks/release.js
@@ -24,7 +24,7 @@ const execaPipeError = (...args) => {
   const [command, ...options] = params;
   const exec = execa(command, options);
 
-  exec.stderr.pipe(process.stderr);
+  // exec.stderr.pipe(process.stderr);
 
   return exec;
 };

--- a/stack/tasks/release.js
+++ b/stack/tasks/release.js
@@ -252,6 +252,7 @@ const confirmedRelease = (type, version) => {
         `));
     }),
     () => execaSeries([
+      'npm run test',
       'npm run build',
       'git add ./dist ./docs',
       'git commit -m"rebuild"',

--- a/stack/tasks/release.js
+++ b/stack/tasks/release.js
@@ -246,7 +246,7 @@ const confirmedRelease = (type, version) => {
 
   let releaseSteps = [
     () => execaSeries([
-      `git checkout ${TRUNK}`,
+      `git checkout ${TRUNK} --quiet`,
       'git pull',
       `git checkout -b ${RELEASE_TMP}`,
     ]).then(() => {

--- a/stack/tasks/release.js
+++ b/stack/tasks/release.js
@@ -28,17 +28,12 @@ const execaPipeError = (...args) => {
 
   return exec
     .then((result) => {
-      const { stdout, stderr, error } = result || {};
-      const hasError = error || (stderr !== stdout && stderr);
-
-      if (hasError) {
-        throw hasError;
-      }
+      console.log(`>>> resolved | ${command} ${options.join(' ')}`);
 
       return result;
     })
     .catch((reason) => {
-      console.error(reason);
+      console.log(`>>> rejected | ${command} ${options.join(' ')}`);
 
       throw reason;
     });

--- a/stack/tasks/release.js
+++ b/stack/tasks/release.js
@@ -18,10 +18,15 @@ const BETA = 'beta';
 
 process.stdin.setEncoding('utf8');
 
-const execaPipeError = (...args) => {
-  const isCommand = args.length === 1;
-  const params = isCommand ? args[0].split(/\s+/) : args;
+const execaPipeError = (file, ...args) => {
+  if (typeof file !== 'string') {
+    return Promise.reoslve();
+  }
+
+  const isCommand = args.length === 0;
+  const params = isCommand ? file.split(/\s+/) : [file, ...args];
   const [command, ...options] = params;
+
   const exec = execa(command, options);
 
   return exec
@@ -236,6 +241,8 @@ const confirmedRelease = (type, version) => {
 
   const isHotfix = type === HOTFIX;
   const TRUNK = isHotfix ? MASTER_TRUNK : DEVELOP_TRUNK;
+  const { scripts } = pkj;
+  const hasTestScript = scripts.test;
 
   let releaseSteps = [
     () => execaSeries([
@@ -248,7 +255,7 @@ const confirmedRelease = (type, version) => {
         `));
     }),
     () => execaSeries([
-      // 'npm run test',
+      hasTestScript && 'npm run test',
       'npm run build',
       'git add ./dist ./docs',
       'git commit -m"rebuild"',

--- a/stack/tasks/release.js
+++ b/stack/tasks/release.js
@@ -28,8 +28,11 @@ const execaPipeError = (...args) => {
 
   return exec
     .then((result) => {
-      if (result && result.error) {
-        throw result.error;
+      const { stderr, error } = result || {};
+      const hasError = error || stderr;
+
+      if (hasError) {
+        throw hasError;
       }
 
       return result;

--- a/stack/tasks/release.js
+++ b/stack/tasks/release.js
@@ -26,7 +26,11 @@ const execaPipeError = (...args) => {
 
   // exec.stderr.pipe(process.stderr);
 
-  return exec;
+  return exec.catch((reason) => {
+    console.error(reason);
+
+    throw reason;
+  });
 };
 
 const execaSeries = args => promiseSeries(args.map(arg => () => execaPipeError(arg)));

--- a/stack/tasks/release.js
+++ b/stack/tasks/release.js
@@ -255,7 +255,7 @@ const confirmedRelease = (type, version) => {
         `));
     }),
     () => execaSeries([
-      'npm run test',
+      // 'npm run test',
       'npm run build',
       'git add ./dist ./docs',
       'git commit -m"rebuild"',

--- a/stack/tasks/release.js
+++ b/stack/tasks/release.js
@@ -28,8 +28,8 @@ const execaPipeError = (...args) => {
 
   return exec
     .then((result) => {
-      const { stderr, error } = result || {};
-      const hasError = error || stderr;
+      const { stdout, stderr, error } = result || {};
+      const hasError = error || (stderr !== stdout && stderr);
 
       if (hasError) {
         throw hasError;

--- a/stack/tasks/release.js
+++ b/stack/tasks/release.js
@@ -24,8 +24,6 @@ const execaPipeError = (...args) => {
   const [command, ...options] = params;
   const exec = execa(command, options);
 
-  // exec.stderr.pipe(process.stderr);
-
   return exec
     .then((result) => {
       console.log(`>>> resolved | ${command} ${options.join(' ')}`);

--- a/stack/tasks/release.js
+++ b/stack/tasks/release.js
@@ -248,7 +248,7 @@ const confirmedRelease = (type, version) => {
     () => execaSeries([
       `git checkout ${TRUNK} --quiet`,
       'git pull',
-      `git checkout -b ${RELEASE_TMP}`,
+      `git checkout -b ${RELEASE_TMP}  --quiet`,
     ]).then(() => {
       console.log(chalk.cyan(outdent`
           Step 1 complete...
@@ -273,7 +273,7 @@ const confirmedRelease = (type, version) => {
     }),
     () => execaSeries([
       `npm publish ${version === BETA ? ' --tag beta' : ''}`,
-      `git checkout ${TRUNK}`,
+      `git checkout ${TRUNK} --quiet`,
       `git merge --ff-only ${RELEASE_TMP}`,
       'git push',
       'git push --tags',
@@ -288,7 +288,7 @@ const confirmedRelease = (type, version) => {
     releaseSteps = [
       ...releaseSteps,
       () => execaSeries([
-        `git checkout ${MASTER_TRUNK}`,
+        `git checkout ${MASTER_TRUNK} --quiet`,
         `git merge --no-ff ${DEVELOP_TRUNK}`,
         'git push',
         'git push --tags',
@@ -298,7 +298,7 @@ const confirmedRelease = (type, version) => {
           `));
       }),
       () => execaSeries([
-        `git checkout ${DEVELOP_TRUNK}`,
+        `git checkout ${DEVELOP_TRUNK} --quiet`,
         `git merge --ff-only ${MASTER_TRUNK}`,
         'git push',
         'git push --tags',

--- a/stack/tasks/release.js
+++ b/stack/tasks/release.js
@@ -26,11 +26,19 @@ const execaPipeError = (...args) => {
 
   // exec.stderr.pipe(process.stderr);
 
-  return exec.catch((reason) => {
-    console.error(reason);
+  return exec
+    .then((result) => {
+      if (result && result.error) {
+        throw result.error;
+      }
 
-    throw reason;
-  });
+      return result;
+    })
+    .catch((reason) => {
+      console.error(reason);
+
+      throw reason;
+    });
 };
 
 const execaSeries = args => promiseSeries(args.map(arg => () => execaPipeError(arg)));

--- a/stack/tasks/release.js
+++ b/stack/tasks/release.js
@@ -19,10 +19,6 @@ const BETA = 'beta';
 process.stdin.setEncoding('utf8');
 
 const execaPipeError = (file, ...args) => {
-  if (typeof file !== 'string') {
-    return Promise.reoslve();
-  }
-
   const isCommand = args.length === 0;
   const params = isCommand ? file.split(/\s+/) : [file, ...args];
   const [command, ...options] = params;
@@ -42,7 +38,9 @@ const execaPipeError = (file, ...args) => {
     });
 };
 
-const execaSeries = args => promiseSeries(args.map(arg => () => execaPipeError(arg)));
+const execaSeries = args => promiseSeries(args
+  .filter(arg => typeof arg === 'string' && arg.length)
+  .map(arg => () => execaPipeError(arg)));
 
 console.log(chalk.cyan(outdent`
 

